rust/lancedb/src/embeddings/openai.rs

```diff
@@ -0,0 +1,262 @@
+            _ => Err(Error::InvalidInput {
+                message: "Invalid input. Available models are: 'text-embedding-3-small', 'text-embedding-ada-002', 'text-embedding-3-large' ".to_string()
+            }),
+        }
+    }
+}
+
+impl ToString for EmbeddingModel {
+    fn to_string(&self) -> String {
+        match self {
+            Self::TextEmbedding3Small => "text-embedding-3-small".to_string(),
+            Self::TextEmbeddingAda002 => "text-embedding-ada-002".to_string(),
+            Self::TextEmbedding3Large => "text-embedding-3-large".to_string(),
+        }
+    }
+}
+
+impl TryFrom<&str> for EmbeddingModel {
+    type Error = Error;
+
+    fn try_from(value: &str) -> std::result::Result<Self, Self::Error> {
+        value.parse()
+    }
+}
+
+pub struct OpenAIEmbeddingFunction {
+    source_column: String,
+    model: EmbeddingModel,
+    api_key: String,
+    api_base: Option<String>,
+    org_id: Option<String>,
+}
+
+impl std::fmt::Debug for OpenAIEmbeddingFunction {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        // let's be safe and not print the full API key
+        let creds_display = if self.api_key.len() > 6 {
+            format!(
+                "{}***{}",
+                &self.api_key[0..2],
+                &self.api_key[self.api_key.len() - 4..]
+            )
+        } else {
+            "[INVALID]".to_string()
+        };
+
+        f.debug_struct("OpenAI")
+            .field("source_column", &self.source_column)
+            .field("model", &self.model)
+            .field("credentials", &creds_display)
+            .field("api_base", &self.api_base)
+            .field("org_id", &self.org_id)
+            .finish()
+    }
+}
+
+impl OpenAIEmbeddingFunction {
+    pub fn new<S: Into<String>, A: Into<String>>(source_column: S, api_key: A) -> Self {
+        Self::new_impl(
+            source_column.into(),
+            api_key.into(),
+            EmbeddingModel::TextEmbeddingAda002,
+        )
+    }
+
+    pub fn new_with_model<S: Into<String>, A: Into<String>, M: TryInto<EmbeddingModel>>(
+        source_column: S,
+        api_key: A,
+        model: M,
+    ) -> crate::Result<Self>
+    where
+        M::Error: Into<crate::Error>,
+    {
+        Ok(Self::new_impl(
+            source_column.into(),
+            api_key.into(),
+            model.try_into().map_err(|e| e.into())?,
+        ))
+    }
+
+    /// concrete implementation to reduce monomorphization
+    fn new_impl(source_column: String, api_key: String, model: EmbeddingModel) -> Self {
+        Self {
+            source_column,
+            model,
+            api_key,
+            api_base: None,
+            org_id: None,
+        }
+    }
+    pub fn api_base<S: Into<String>>(mut self, api_base: S) -> Self {
+        self.api_base = Some(api_base.into());
+        self
+    }
+
+    pub fn org_id<S: Into<String>>(mut self, org_id: S) -> Self {
+        self.org_id = Some(org_id.into());
+        self
+    }
```

Comment:
Doesn't have to be this PR but can we document these public methods at some point?  I don't know what `api_base` or `org_id` are referring to.